### PR TITLE
[Sol2Yul] Fixed an ICE on struct member copy

### LINKED ISSUE #12220: [Sol->Yul] ICE when assigning msg.data to a struct member
```
struct St0 {
  bytes el0;
}
contract C {
  function f() external {
    St0 memory x;
    x.el0 = msg.data;
  }
}
```

throws

https://github.com/ethereum/solidity/blob/5eb97fa6ee059022ac2fe146649a57b94a8572fc/libsolidity/codegen/ir/IRGeneratorForStatements.cpp#L2918

To repro

```
$ solc --experimental-via-ir test.sol
```

### LINKED ISSUE #12558: ICE when copying storage structs to memory arrays via IR
## Description
Found while working on the external test for [Pool Together v4](https://github.com/pooltogether/v4-core). The contract does not compile via IR at all due to an Internal Compiler Error. With legacy codegen the compilation succeeds.

## Steps to Reproduce
Here's a distilled repro for the ICE:
```solidity
struct S {
    uint x;
}

contract C {
    S sStorage;

    function f() external {
        S[] memory sMemory;

        sMemory[0] = sStorage;
    }
}
```
```
Internal compiler error:
/solidity/libsolidity/codegen/ir/IRGeneratorForStatements.cpp(2962): Throw in function solidity::frontend::IRGeneratorForStatements::writeToLValue(const solidity::frontend::IRLValue&, const solidity::frontend::IRVariable&)::<lambda(const solidity::frontend::IRLValue::Memory&)>
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Solidity assertion failed
[solidity::util::tag_comment*] = Solidity assertion failed
```

It only happens when the target struct is a part of an array (does not matter whether dynamic or static). Does not happen when the target is just a single struct.

**EDIT**: When this is fixed, remember to re-enable the `ir-optimize-evm+yul` preset in the `pool-together` external test (which is now in `develop`): https://github.com/ethereum/solidity/blob/93f7d5f8f0be28d84158a6ccde7fd7d2050598d6/test/externalTests/pool-together.sh#L48

## Environment

- Compiler version: 0.8.11

### PR DESCRIPTION
**Pull request by hrkrshnn**
Closes https://github.com/ethereum/solidity/issues/12220
Closes https://github.com/ethereum/solidity/issues/12558

### CODE REVIEW

**Review by ekpyron — COMMENTED** [review]
Mainly needs a changelog entry.

**Comment by hrkrshnn** [comment]
@ekpyron Does it need a changelog entry? It's a change in Yul.

**Comment by ekpyron** [comment]
> @ekpyron Does it need a changelog entry? It's a change in Yul.

Yes, but those have been added to the changelog for a while now.

**Comment by ekpyron** [comment]
So, can you add the entry ;-)?

File: Changelog.md [hunk 1]
```diff
@@ -16,7 +16,8 @@ Bugfixes:
  * Code Generator: Fix ICE when doing an explicit conversion from ``string calldata`` to ``bytes``.
  * Control Flow Graph: Perform proper virtual lookup for modifiers for uninitialized variable and unreachable code analysis.
  * Immutables: Fix wrong error when the constructor of a base contract uses ``return`` and the parent contract contains immutable variables.
- * IR Generator: Fix IR syntax error when copying storage arrays of structs containing functions.
+  * IR Generator: Fix IR syntax error when copying storage arrays of structs containing functions.
```
Comment (ekpyron) [inline Changelog.md]: Weird... why is this part of the diff? Looks identical to me.

Comment (hrkrshnn) [reply Changelog.md]: Ah sorry. An extra space by accident. Fixed.

File: Changelog.md [hunk 2]
```diff
@@ -17,6 +17,7 @@ Bugfixes:
  * Control Flow Graph: Perform proper virtual lookup for modifiers for uninitialized variable and unreachable code analysis.
  * Immutables: Fix wrong error when the constructor of a base contract uses ``return`` and the parent contract contains immutable variables.
  * IR Generator: Fix IR syntax error when copying storage arrays of structs containing functions.
+ * IR Generator: Fix IR syntax error when copying reference types in calldata and storage to struct members in memory.
```
Comment (ekpyron) [inline Changelog.md]: It's not only struct members, but also array elements, isn't it?
Might actually be worth another simple test case.

Comment (hrkrshnn) [reply Changelog.md]: Actually yeah, also array members, the test `test/libsolidity/semanticTests/structs/copy_from_calldata.sol` shows it.

Comment (hrkrshnn) [inline Changelog.md]:
```suggestion
 * IR Generator: Fix IR syntax error when copying reference types in calldata and storage to struct or array members in memory.
```
@ekpyron would that work? I can commit and squash it afterwards.

Comment (hrkrshnn) [reply Changelog.md]: Hmm. Github is not rendering it properly! I'll add "or array" members.

Comment (ekpyron) [reply Changelog.md]: We could add a few more tests, i.e. copying storage arrays to a memory array element, etc.... we can also consider that part of https://github.com/ethereum/solidity/issues/12559, but we can also do it while we're at it...

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -2982,8 +2982,11 @@ void IRGeneratorForStatements::writeToLValue(IRLValue const& _lvalue, IRVariable
 				{
 					solAssert(_lvalue.type.sizeOnStack() == 1);
 					auto const* valueReferenceType = dynamic_cast<ReferenceType const*>(&_value.type());
-					solAssert(valueReferenceType && valueReferenceType->dataStoredIn(DataLocation::Memory));
-					appendCode() << "mstore(" + _memory.address + ", " + _value.part("mpos").name() + ")\n";
+					solAssert(valueReferenceType);
+					if (valueReferenceType->dataStoredIn(DataLocation::Memory))
+						appendCode() << "mstore(" + _memory.address + ", " + _value.part("mpos").name() + ")\n";
+					else
+						appendCode() << "mstore(" + _memory.address + ", " + m_utils.conversionFunction(_value.type(), _lvalue.type) + "(" + _value.commaSeparatedList() + "))\n";
```
Comment (bshastry) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:2992]: Does the conversion function copy storage/calldata value to memory and return its location?

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:2992]: Pretty much, yep!

**Review by bshastry — DISMISSED** [review]
Looks good to me.

**Comment by ekpyron** [comment]
Changelog still needs to include the array part.

**Comment by hrkrshnn** [comment]
@ekpyron sorry, will do this today.

**Comment by ekpyron** [comment]
Still missing the changelog entry change ;-).

File: test/libsolidity/semanticTests/structs/copy_struct_array_from_storage.sol [hunk 4]
```diff
@@ -0,0 +1,96 @@
+pragma abicoder v2;
+
```
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/structs/copy_struct_array_from_storage.sol:2]: Added this example: @ekpyron

Comment (hrkrshnn) [reply test/libsolidity/semanticTests/structs/copy_struct_array_from_storage.sol:2]: But this one surprisingly also compiles in other versions. So independent of the fix.

Comment (ekpyron) [reply test/libsolidity/semanticTests/structs/copy_struct_array_from_storage.sol:2]: Hm... I'm pretty sure that there were cases that didn't work in other versions... let me see...

Comment (ekpyron) [reply test/libsolidity/semanticTests/structs/copy_struct_array_from_storage.sol:2]:
```
contract C {
        uint256[] x;

        function f() public {
                uint256[][] memory y = new uint256[][](1);
                y[0] = x;
        }
}
```
crashes on 0.8.11 with ``--experimental-via-ir --bin`` for me...

Comment (ekpyron) [reply test/libsolidity/semanticTests/structs/copy_struct_array_from_storage.sol:2]: Actually, this test case you added also does, so it's fine - but it's *not* independent of the fix :-).

Comment (hrkrshnn) [reply test/libsolidity/semanticTests/structs/copy_struct_array_from_storage.sol:2]: Ah, my bad. Forgot to add `--ir` flag to the test. So yeah, it does crash!

File: Changelog.md [hunk 5]
```diff
@@ -17,6 +17,7 @@ Bugfixes:
  * Control Flow Graph: Perform proper virtual lookup for modifiers for uninitialized variable and unreachable code analysis.
  * Immutables: Fix wrong error when the constructor of a base contract uses ``return`` and the parent contract contains immutable variables.
  * IR Generator: Fix IR syntax error when copying storage arrays of structs containing functions.
+ * IR Generator: Fix IR syntax error when copying reference types in calldata and storage to struct or array members in memory.
```
Comment (ekpyron) [inline Changelog.md]: Actually it wasn't really an IR syntax error, but a failing assertion, resp. it was unimplemented so far...

**Review by ekpyron — DISMISSED** [review]
Code change and test look good, the changelog entry is still not really accurate, but I'm already approving.

File: Changelog.md (hunk 5 above)
Comment (hrkrshnn) [reply Changelog.md]: Fixing this now.

File: Changelog.md [hunk 6]
```diff
@@ -17,6 +17,7 @@ Bugfixes:
  * Control Flow Graph: Perform proper virtual lookup for modifiers for uninitialized variable and unreachable code analysis.
  * Immutables: Fix wrong error when the constructor of a base contract uses ``return`` and the parent contract contains immutable variables.
  * IR Generator: Fix IR syntax error when copying storage arrays of structs containing functions.
+ * IR Generator: Fix ICE when copying reference types in calldata and storage to struct or array members in memory.
```
Comment (hrkrshnn) [inline Changelog.md]: @ekpyron Changed to ICE.

**Review by ekpyron — DISMISSED** [review]

File: Changelog.md (hunk 6 above)
Comment (ekpyron) [reply Changelog.md]: Actually, now it isn't sorted alphabetically anymore ;-D.

Comment (hrkrshnn) [reply Changelog.md]: Fixed!

**Comment by ekpyron** [comment]
It has a merge conflict now and something weird happened to CI (but looks like a weird CirecleCI issue independent of the PR) - so I'd hope a rebase will fix it...

**Comment by Assawal** [comment]
Sent from my Huawei phone-------- Original message --------From: Daniel Kirchner ***@***.***>Date: Mon, 31 Jan 2022, 18:37To: ethereum/solidity ***@***.***>Cc: Subscribed ***@***.***>Subject: Re: [ethereum/solidity] [Sol2Yul] Fixed an ICE on struct member copy (PR #12579)
It has a merge conflict now and something weird happened to CI (but looks like a weird CirecleCI issue independent of the PR) - so I'd hope a rebase will fix it...

**Comment by hrkrshnn** [comment]
Rebased @ekpyron :)

**Review by ekpyron — DISMISSED** [review]

File: Changelog.md [hunk 7]
```diff
@@ -18,6 +18,7 @@ Bugfixes:
  * Immutables: Fix wrong error when the constructor of a base contract uses ``return`` and the parent contract contains immutable variables.
  * IR Generator: Add missing cleanup during the conversion of fixed bytes types to smaller fixed bytes types.
  * IR Generator: Add missing cleanup for indexed event arguments of value type.
+ * IR Generator: Fix ICE when copying reference types in calldata and storage to struct or array members in memory.
```
Comment (chriseth) [inline Changelog.md]: We use ICE liberally, but the term is really difficult to understand without background and "internal error" is almost as short.

Comment (hrkrshnn) [reply Changelog.md]: I can fix this. There are liked 4 other ICEs in the same section! Will leave them untouched in this PR.

Comment (hrkrshnn) [reply Changelog.md]: Fixed!

**Review by ekpyron — APPROVED** [review]

**Pull request merged**